File: 2019/python/day2.py
import copy
import itertools


def load_program(s):
    return dict(enumerate(map(int, s.split(','))))


def draw(tape):
    n = max(tape.keys()) + 1
    return ",".join(str(tape.get(i)) for i in range(n))


def run_and_draw(s):
    return draw(run(load_program(s))[1])


def run(tape):
    pc = 0
    while True:
        op = tape[pc]
        if op == 99:
            return pc, tape
        arg_1, arg_2, arg_3 = tape[pc + 1], tape[pc + 2], tape[pc + 3]
        op_1 = tape.get(arg_1, 0)
        op_2 = tape.get(arg_2, 0)
        if op == 1:
            tape[arg_3] = op_1 + op_2
        elif op == 2:
            tape[arg_3] = op_1 * op_2
        else:
            raise Exception(f"invalid opcode {op} (pc = {pc})")
        pc += 4


def main():
    # print(draw(load_program("1,0,0,0,99")))
    assert run_and_draw("1,0,0,0,99") == "2,0,0,0,99"
    assert run_and_draw("2,4,4,5,99,0") == "2,4,4,5,99,9801"
    assert run_and_draw("1,1,1,4,99,5,6,0,99") == "30,1,1,4,2,5,6,0,99"
    src = None
    with open('assets/day2-input') as f:
        src = load_program(f.readline().strip())
        tape = copy.copy(src)
        tape[1] = 12
        tape[2] = 2
        print("part 1:", run(tape)[1][0])

    for verb, noun in itertools.product(range(0, 100), range(0, 100)):
        tape = copy.copy(src)
        tape[1] = verb
        tape[2] = noun
        if run(tape)[1][0] == 19690720:
            print("part 2:", 100 * verb + noun)


if __name__ == '__main__':
    main()

File: 2019/rust/src/day2.rs
use super::intcode;
use std::fs::File;
use std::io;
use std::io::Read;

pub fn day2(input: &str) -> io::Result<()> {
    let mut f = File::open(input)?;
    let mut tape_str = String::new();
    f.read_to_string(&mut tape_str)?;
    let mut cpu = intcode::IntCpu::from_str(tape_str.trim()).unwrap();
    cpu.tape_write(1, 12).unwrap();
    cpu.tape_write(2, 2).unwrap();

    println!(
        "part 1: {}",
        cpu.run().unwrap_or_else(|| {
            cpu.dump();
            0
        })
    );

    for verb in 0..100 {
        for noun in 0..100 {
            cpu.reset();
            cpu.tape_write(1, verb).unwrap();
            cpu.tape_write(2, noun).unwrap();
            if cpu.run().unwrap() == 19690720 {
                println!("part 2: {}", 100 * verb + noun)
            }
        }
    }

    Ok(())
}
